pkg/services/ngalert/state/manager.go

```diff
@@ -65,12 +62,10 @@ func (st *Manager) Close() {
 	st.quit <- struct{}{}
 }
 
-func (st *Manager) Warm(ctx context.Context) {
+func (st *Manager) Warm(ctx context.Context, rulesReader RuleReader) {
```

Comment:
We can optimize it even further and make a cached reader (fetcher in scheduler), reuse it in the scheduler and state manager. This way we do not need to fetch all rules during the startup twice

Reply:
Do we even need to pass in a `RuleReader`?  What if we passed in `[]*models.AlertRule`?

Reply:
Yes, we can do that! Actually, I did that by introducing Warm method in the scheduler that returns the slice of rules but then discarded changes as it was too many changes for a quick refactor.